modelling/src/main/java/org/axonframework/modelling/command/StatefulCommandHandlingComponent.java

```diff
@@ -0,0 +1,308 @@
+        return handlingComponent.supportedCommands();
+    }
+
+    @Override
+    public <ID, M> StatefulCommandHandlingComponent registerModel(
+            String name,
+            Class<M> modelClass,
+            Function<CommandMessage<?>, ID> commandIdResolver,
+            ModelLoader<ID, M> loadFunction) {
+        if (modelDefinitions.stream().anyMatch(md -> md.name().equals(name) && md.modelClass().equals(modelClass))) {
+            throw new IllegalStateException("Model with name " + name + " already registered");
+        }
+        modelDefinitions.add(new ModelDefinition<>(name, modelClass, commandIdResolver, loadFunction));
+        return this;
+    }
+
+    @Override
+    public void describeTo(ComponentDescriptor descriptor) {
+        descriptor.describeProperty("name", name);
+        descriptor.describeProperty("loadModelsEagerly", loadModelsEagerly);
+        descriptor.describeProperty("handlingComponent", handlingComponent);
+        descriptor.describeProperty("modelDefinitions", modelDefinitions);
+    }
+
+    /**
+     * A definition of a model that can be registered. Used for resolving of the models during execution of the stateful
+     * command handlers that are subscribed to this component. Solely used for internal purposes.
+     */
+    private record ModelDefinition<ID, M>(
+            String name,
+            Class<M> modelClass,
+            Function<CommandMessage<?>, ID> commandIdResolver,
+            ModelLoader<ID, M> loader
+    ) {
+
+    }
+
+    /**
+     * A container for models that either loads models lazily or eagerly. Used during the handling of commands by the
+     * stateful command handlers.
+     * <p>
+     * Any defined model will only be initialized once per command handling process.
+     */
+    private class StatefulCommandModelModelContainer implements ModelContainer {
+
+        private final CommandMessage<?> command;
+        private final ProcessingContext context;
+        private final List<LoadedModelDefinition<?, ?>> loadedModels = new CopyOnWriteArrayList<>();
+
+        /**
+         * Initializes a new {@link ModelContainer} for the given {@code command} and {@code context} that lazily loads
+         * the models.
+         *
+         * @param command The command to load models for
+         * @param context The context in which the command is handled
+         */
+        private StatefulCommandModelModelContainer(CommandMessage<?> command, ProcessingContext context,
+                                                   boolean loadEagerly) {
+            this.command = command;
+            this.context = context;
+
+            if (loadEagerly) {
+                modelDefinitions.forEach(this::eagerLoadModelIfHasIdentifier);
+            }
+        }
+
+        @Override
+        public <T> T modelOf(@Nonnull Class<T> modelType, String name) {
+            T loadedModel = alreadyLoadedModelOf(modelType, name);
+            if (loadedModel != null) {
```

Comment:
I think this should be documented. Someone might expect it to load the updated model on the second invocation if something changes in the underlying storage between invocations (or without using AxonServer, they might expect that transaction level has an influence on that).

Reply:
I have improved the javadoc in this regard.

However, I think it's expected to have some caching on ProcessingContext level. 
Without caching the implementation would be a lot simpler though :) Depending on what Steven says, we can remove it.